File: frontend/src/components/benchmark/content/preferential_attachment/implementations/preferential_attachment.js
class PreferentialAttachment {
    constructor() {
        this.adjacency = new Map();
        this.nodes = new Set();
    }

    addEdge(u, v) {
        if (!this.adjacency.has(u)) {
            this.adjacency.set(u, new Set());
        }
        if (!this.adjacency.has(v)) {
            this.adjacency.set(v, new Set());
        }
        this.adjacency.get(u).add(v);
        this.adjacency.get(v).add(u);
        this.nodes.add(u);
        this.nodes.add(v);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    preferentialAttachmentScore(u, v) {
        if (!this.adjacency.has(u) || !this.adjacency.has(v)) {
            return 0.0;
        }
        return this.adjacency.get(u).size * this.adjacency.get(v).size;
    }

    scoreAllPairs() {
        const scores = [];
        const nodeList = Array.from(this.nodes);

        for (let i = 0; i < nodeList.length; i++) {
            for (let j = i + 1; j < nodeList.length; j++) {
                const u = nodeList[i];
                const v = nodeList[j];
                const score = this.preferentialAttachmentScore(u, v);
                if (score > 0) {
                    scores.push({ pair: [u, v], score: score });
                }
            }
        }

        scores.sort((a, b) => b.score - a.score);
        return scores;
    }

    topKPredictions(k = 10) {
        const scores = this.scoreAllPairs();
        return scores.slice(0, k);
    }
}

module.exports = PreferentialAttachment;

File: frontend/src/components/benchmark/content/preferential_attachment/implementations/preferential_attachment.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

class PreferentialAttachment {
private:
    unordered_map<int, unordered_set<int>> adjacency;
    unordered_set<int> nodes;

public:
    void add_edge(int u, int v) {
        adjacency[u].insert(v);
        adjacency[v].insert(u);
        nodes.insert(u);
        nodes.insert(v);
    }

    void build_from_edges(const vector<pair<int, int>>& edges) {
        for (const auto& edge : edges) {
            add_edge(edge.first, edge.second);
        }
    }

    double preferential_attachment_score(int u, int v) {
        if (adjacency.find(u) == adjacency.end() || adjacency.find(v) == adjacency.end()) {
            return 0.0;
        }
        return (double)adjacency[u].size() * adjacency[v].size();
    }

    vector<pair<pair<int, int>, double>> score_all_pairs() {
        vector<pair<pair<int, int>, double>> scores;
        vector<int> node_list(nodes.begin(), nodes.end());

        for (size_t i = 0; i < node_list.size(); i++) {
            for (size_t j = i + 1; j < node_list.size(); j++) {
                int u = node_list[i];
                int v = node_list[j];
                double score = preferential_attachment_score(u, v);
                if (score > 0) {
                    scores.push_back({{u, v}, score});
                }
            }
        }

        sort(scores.begin(), scores.end(),
             [](const auto& a, const auto& b) {
                 return a.second > b.second;
             });

        return scores;
    }

    vector<pair<pair<int, int>, double>> top_k_predictions(int k = 10) {
        auto scores = score_all_pairs();
        vector<pair<pair<int, int>, double>> result;
        for (int i = 0; i < min(k, (int)scores.size()); i++) {
            result.push_back(scores[i]);
        }
        return result;
    }
};